openstack/client.go

```diff
@@ -212,125 +215,81 @@ func NewIdentityV3(client *gophercloud.ProviderClient, eo gophercloud.EndpointOp
 	return &gophercloud.ServiceClient{
 		ProviderClient: client,
 		Endpoint:       endpoint,
+		Type:           clientType,
 	}, nil
 }
 
-// NewObjectStorageV1 creates a ServiceClient that may be used with the v1 object storage package.
-func NewObjectStorageV1(client *gophercloud.ProviderClient, eo gophercloud.EndpointOpts) (*gophercloud.ServiceClient, error) {
-	eo.ApplyDefaults("object-store")
+func initClientOpts(client *gophercloud.ProviderClient, eo gophercloud.EndpointOpts, clientType string) (*gophercloud.ServiceClient, error) {
+	sc := new(gophercloud.ServiceClient)
+	eo.ApplyDefaults(clientType)
 	url, err := client.EndpointLocator(eo)
 	if err != nil {
-		return nil, err
+		return sc, err
 	}
-	return &gophercloud.ServiceClient{ProviderClient: client, Endpoint: url}, nil
+	sc.ProviderClient = client
+	sc.Endpoint = url
+	sc.Type = clientType
```

Comment:
Currently, there is no mechanism for returning information from endpoint that would be easily extensible. That's why I'm proposing to add a struct as a new endpoint return value. The struct can be easily extended in the future and doesn't have to contain all currently available information.

As you introduced the `func initClientOpts`, the `client.EndpointLocator` is called at less places than before. So it's now easier to add new return values to the `client.EndpointLocator` function.

I agree with you solution.